syntax = "proto3";
package chat;

service ChatService {
    rpc GetAllUsers (GetAllUsersRequest) returns (GetAllUsersReply);
    rpc RecieveMessages (RecieveMessagesRequest) returns (stream RecieveMessagesReply);
    rpc SendMessage (SendMessageRequest) returns (SendMessageReply);
    rpc RegisterUser (RegisterUserRequest) returns (RegisterUserReply);
    rpc LoginUser (LoginUserRequest) returns (LoginUserReply);
}

//-------------------------------------//

message GetAllUsersRequest {
}

message GetAllUsersReply {
    repeated UserInfo users = 1;
}
//-------------------------------------//
message RecieveMessagesRequest {
    string to_user_login = 1;
}

message RecieveMessagesReply {
    Message message = 1;
}
//-------------------------------------//
message SendMessageRequest {
    Message message = 1;
}

message SendMessageReply {
}
//-------------------------------------//

message UserInfo {
    string login = 1;
    string full_name = 2;
}

message MessageBody {
    string body = 1;
    string timestamp = 2;
}

message Message {
    string from_user_login = 1;
    string to_user_login = 2;
    MessageBody body = 3;
}

message EtcdUserInfo {
    UserInfo user_info = 1;
    bool is_active = 2;
    string hashed_password = 3;
    string register_timestamp = 4;
}
//-------------------------------------//

message LoginUserRequest {
    string login = 1;
    string password = 2;
}

message LoginUserReply {
}
//-------------------------------------//

message RegisterUserRequest {
    UserInfo user_info = 1;
    string password = 2;
}

message RegisterUserReply {
}
